File: trunk/NowPlaying/ObjectiveC/Classes/LocationUtilities.m
// Copyright (C) 2008 Cyrus Najmabadi
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the Free
// Software Foundation; either version 2 of the License, or (at your option) any
// later version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
// details.
//
// You should have received a copy of the GNU General Public License along with
// this program; if not, write to the Free Software Foundation, Inc., 51
// Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

#import "LocationUtilities.h"

#import "NetworkUtilities.h"
#import "XmlElement.h"

@implementation LocationUtilities

+ (NSString*) findUSPostalCode:(CLLocation*) location {
    CLLocationCoordinate2D coordinates = location.coordinate;
    double latitude = coordinates.latitude;
    double longitude = coordinates.longitude;
    NSString* url = [NSString stringWithFormat:@"http://ws.geonames.org/findNearbyPostalCodes?lat=%f&lng=%f&maxRows=1", latitude, longitude];

    XmlElement* geonamesElement = [NetworkUtilities xmlWithContentsOfAddress:url important:YES];
    XmlElement* codeElement = [geonamesElement element:@"code"];
    XmlElement* postalElement = [codeElement element:@"postalcode"];
    XmlElement* countryElement = [codeElement element:@"countryCode"];

    if ([@"CA" isEqual:countryElement.text]) {
        return nil;
    }

    return postalElement.text;
}


+ (NSString*) findCAPostalCode:(CLLocation*) location {
    CLLocationCoordinate2D coordinates = location.coordinate;
    double latitude = coordinates.latitude;
    double longitude = coordinates.longitude;
    NSString* url = [NSString stringWithFormat:@"http://geocoder.ca/?latt=%f&longt=%f&geoit=xml&reverse=Reverse+GeoCode+it", latitude, longitude];

    XmlElement* geodataElement = [NetworkUtilities xmlWithContentsOfAddress:url
                                                                  important:YES];
    XmlElement* postalElement = [geodataElement element:@"postal"];
    return postalElement.text;
}


+ (NSString*) findPostalCode:(CLLocation*) location {
    NSString* postalCode = [self findUSPostalCode:location];
    if (postalCode == nil) {
        postalCode = [self findCAPostalCode:location];
    }
    return postalCode;
}

@end

File: trunk/NowPlaying/Android/src/org/metasyntactic/utilities/LocationUtilities.java
//Copyright 2008 Cyrus Najmabadi
//
//Licensed under the Apache License, Version 2.0 (the "License");
//you may not use this file except in compliance with the License.
//You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
//Unless required by applicable law or agreed to in writing, software
//distributed under the License is distributed on an "AS IS" BASIS,
//WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//See the License for the specific language governing permissions and
//limitations under the License.
package org.metasyntactic.utilities;

import org.metasyntactic.data.Location;
import static org.metasyntactic.utilities.StringUtilities.isNullOrEmpty;
import static org.metasyntactic.utilities.XmlUtilities.element;
import static org.metasyntactic.utilities.XmlUtilities.text;
import org.w3c.dom.Element;

public class LocationUtilities {
  private LocationUtilities() {
  }

  private static Location findLocationWithGeonames(final double latitude, final double longitude) {
    final String url = "http://ws.geonames.org/findNearbyPostalCodes?lat=" + latitude + "&lng=" + longitude + "&maxRows=1";

    final Element geonamesElement = NetworkUtilities.downloadXml(url, true);
    final Element codeElement = element(geonamesElement, "code");
    final String postalCode = text(element(codeElement, "postalcode"));
    final String country = text(element(codeElement, "countryCode"));

    if (isNullOrEmpty(postalCode)) {
      return null;
    }

    if ("CA".equals(country)) {
      return null;
    }

    final String city = text(element(codeElement, "adminName1"));
    final String state = text(element(codeElement, "adminCode1"));

    return new Location(latitude, longitude, "", city, state, postalCode, country);
  }

  private static Location findLocationWithGeocoder(final double latitude, final double longitude) {
    final String url = "http://geocoder.ca/?latt=" + latitude + "&longt=" + longitude + "&geoit=xml&reverse=Reverse+GeoCode+it";

    final Element geodataElement = NetworkUtilities.downloadXml(url, true);
    final String postalCode = text(element(geodataElement, "postal"));

    if (isNullOrEmpty(postalCode)) {
      return null;
    }

    final String city = text(element(geodataElement, "city"));
    final String state = text(element(geodataElement, "prov"));

    return new Location(latitude, longitude, "", city, state, postalCode, "CA");
  }

  public static Location findLocation(final double latitude, final double longitude) {
    Location result = findLocationWithGeonames(latitude, longitude);
    if (result == null) {
      result = findLocationWithGeocoder(latitude, longitude);
    }
    return result;
  }
}
